binder/src/main/java/io/grpc/binder/AndroidComponentAddress.java

```diff
@@ -177,13 +190,96 @@ public int hashCode() {
   public boolean equals(Object obj) {
     if (obj instanceof AndroidComponentAddress) {
       AndroidComponentAddress that = (AndroidComponentAddress) obj;
-      return bindIntent.filterEquals(that.bindIntent);
+      return bindIntent.filterEquals(that.bindIntent)
+          && Objects.equal(this.targetUser, that.targetUser);
     }
     return false;
   }
 
   @Override
   public String toString() {
-    return "AndroidComponentAddress[" + bindIntent + "]";
+    StringBuilder builder = new StringBuilder("AndroidComponentAddress[");
+    if (targetUser != null) {
+      builder.append(targetUser);
+      builder.append("@");
+    }
+    builder.append(bindIntent);
+    builder.append("]");
+    return builder.toString();
+  }
+
+  /**
+   * Identifies the Android user in which the bind Intent will be evaluated.
+   *
+   * <p>Returns the {@link UserHandle}, or null which means that the Android user hosting the
+   * current process will be used.
+   */
+  @Nullable
+  public UserHandle getTargetUser() {
+    return targetUser;
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /**
+   * Fluently builds instances of {@link AndroidComponentAddress}.
+   */
+  public static class Builder {
+    Intent bindIntent = new Intent(ApiConstants.ACTION_BIND);
+    UserHandle targetUser;
+
+    /**
+     * See {@link Intent#setPackage(String)}.
+     */
+    public Builder setPackage(String pkg) {
+      bindIntent.setPackage(pkg);
+      return this;
+    }
+
+    /**
+     * See {@link Intent#setComponent(ComponentName)}.
+     */
+    public Builder setComponent(ComponentName component) {
+      bindIntent.setComponent(component);
+      return this;
+    }
+
+    /**
+     * See {@link Intent#addCategory(String)}.
+     */
+    public Builder addCategory(String category) {
+      bindIntent.addCategory(category);
+      return this;
+    }
+
+    /**
+     * See {@link Intent#setAction(String)}.
+     */
+    public Builder setAction(String action) {
+      bindIntent.setAction(action);
+      return this;
+    }
+
+    /**
+     * See {@link Intent#setData(Uri)}.
+     */
+    public Builder setData(Uri data) {
+      bindIntent.setData(data);
+      return this;
+    }
+
+    /**
+     * See {@link AndroidComponentAddress#getTargetUser()}.
+     */
+    public Builder setTargetUser(@Nullable UserHandle targetUser) {
+      this.targetUser = targetUser;
+      return this;
+    }
+
+    public AndroidComponentAddress build() {
+      return new AndroidComponentAddress(bindIntent, targetUser);
```

Comment:
We need to duplicate the intent, to avoid it being modified in the AndroidComponentAddress.

Reply:
oops thanks for noticing. PTAL